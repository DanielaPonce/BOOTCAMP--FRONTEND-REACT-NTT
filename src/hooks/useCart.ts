import { useEffect } from 'react';
import { useCartContext } from '../context/CartContext';
import { Product } from '../models/product.types';
import { Actions } from '../utils/actions';
import { LocalStorageConstants } from '../utils/localStorageConstants';

export const useCart = () => {
	const { state, dispatch } = useCartContext();

	useEffect(() => {
		localStorage.setItem(
			LocalStorageConstants.CartKey,
			JSON.stringify(state)
		);
	}, [state]);

	const addToCart = (product: Product) =>
		dispatch({ type: Actions.ADD_TO_CART_TYPE, payload: product });

	const removeFromCart = (id: number) =>
		dispatch({ type: Actions.REMOVE_FROM_CART_TYPE, payload: id });

	const updateQuantity = (id: number, quantity: number) =>
		dispatch({
			type: Actions.UPDATE_QUANTITY_TYPE,
			payload: { id, quantity }
		});

	const clearCart = () => dispatch({ type: Actions.CLEAR_CART_TYPE });

	return {
		cart: state.cart,
		total: state.total,
		itemCount: state.itemCount,
		addToCart,
		removeFromCart,
		updateQuantity,
		clearCart
	};
};
