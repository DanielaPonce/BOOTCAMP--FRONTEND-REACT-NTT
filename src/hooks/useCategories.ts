import { useEffect, useState } from 'react';
import { fetchAllCategories } from '../api/category.api';
import { Category } from '../models/category.types';

export const useCategories = () => {
	const [categories, setCategories] = useState<Category[]>([]);

	useEffect(() => {
		fetchAllCategories().then((data) => setCategories(data));
	}, []);

	return { categories };
};
