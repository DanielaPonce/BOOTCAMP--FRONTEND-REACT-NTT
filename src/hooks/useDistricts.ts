import { useEffect, useState } from 'react';

interface District {
	id: number;
	name: string;
}

export const useDistricts = () => {
	const [districts, setDistricts] = useState<District[]>([]);

	useEffect(() => {
		const fetchDistricts = async () => {
			const response = await fetch('/src/assets/mock/districts.json');
			const data: District[] = await response.json();
			setDistricts(data);
		};

		fetchDistricts();
	}, []);

	return districts;
};
