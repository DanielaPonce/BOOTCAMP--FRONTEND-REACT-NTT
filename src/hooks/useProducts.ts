import { useEffect, useState } from 'react';
import { fetchAllProducts, fetchProductsByCategory } from '../api/product.api';
import { Product } from '../models/product.types';
import { Constants } from '../utils/constants';

export const useProducts = (category: string, search: string) => {
	const [allProducts, setAllProducts] = useState<Product[]>([]);
	const [filteredProducts, setFilteredProducts] = useState<Product[]>([]);
	const [isLoading, setIsLoading] = useState(true);
	const [currentPage, setCurrentPage] = useState(1);

	const itemsPerPage = 8;

	const loadProducts = async () => {
		setIsLoading(true);
		try {
			let products: Product[] = [];
			if (category === Constants.defaultSelectValue) {
				products = await fetchAllProducts();
			} else {
				products = await fetchProductsByCategory(category);
			}
			setAllProducts(products);
			setFilteredProducts(products);
			setCurrentPage(1);
		} catch (error) {
			console.log('Error al cargar los productos:', error);
		} finally {
			setIsLoading(false);
		}
	};

	useEffect(() => {
		loadProducts();
	}, [category]);

	useEffect(() => {
		const filtered = allProducts.filter((product) =>
			product.title.toLowerCase().includes(search.toLowerCase())
		);
		setFilteredProducts(filtered);
		setCurrentPage(1);
	}, [search, allProducts]);

	const paginatedProducts = filteredProducts.slice(
		(currentPage - 1) * itemsPerPage,
		currentPage * itemsPerPage
	);

	const totalPages = Math.ceil(filteredProducts.length / itemsPerPage);

	return {
		products: paginatedProducts,
		isLoading,
		totalPages,
		currentPage,
		setCurrentPage,
		showPagination: totalPages !== 1
	};
};
