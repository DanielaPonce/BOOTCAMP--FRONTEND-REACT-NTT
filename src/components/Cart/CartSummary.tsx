import { FC } from 'react';
import { useCart } from '../../hooks/useCart';

const CartSummary: FC = () => {
	const { total } = useCart();

	return (
		<div className="cart-summary">
			<h3>Total a pagar: S/ {total.toFixed(2)}</h3>
		</div>
	);
};

export default CartSummary;
