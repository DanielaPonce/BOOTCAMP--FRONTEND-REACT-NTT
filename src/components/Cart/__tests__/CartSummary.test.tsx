// si el contenido no es variable o condicionado, o usa servicios un snapshot basta no es necesario agregarle tantos expect
import React from 'react';
import { render, screen } from '@testing-library/react';
import { jest } from '@jest/globals';
import CartSummary from '../CartSummary';
import { useCart } from '../../../hooks/useCart';

jest.mock('../../../hooks/useCart');

describe('CartSummary Component', () => {
	const mockTotal = 123.45;

	beforeEach(() => {
		(useCart as jest.Mock).mockReturnValue({ total: mockTotal });
	});

	test('renders correctly', () => {
		render(<CartSummary />);

		const totalText = screen.getByText(
			`Total a pagar: S/ ${mockTotal.toFixed(2)}`
		);
		expect(totalText).toBeInTheDocument();
	});

	test('formats the total with two decimal places', () => {
		const customTotal = 987.6;
		(useCart as jest.Mock).mockReturnValue({ total: customTotal });

		render(<CartSummary />);

		const formattedTotal = screen.getByText(
			`Total a pagar: S/ ${customTotal.toFixed(2)}`
		);
		expect(formattedTotal).toBeInTheDocument();
		// es lo mismo que la l'inea 34
		expect(formattedTotal).toHaveTextContent('S/ 987.60');
	});

	test('handles a total of 0 correctly', () => {
		(useCart as jest.Mock).mockReturnValue({ total: 0 });

		render(<CartSummary />);

		const zeroTotal = screen.getByText('Total a pagar: S/ 0.00');
		expect(zeroTotal).toBeInTheDocument();
	});

	test('handles negative totals', () => {
		(useCart as jest.Mock).mockReturnValue({ total: -50.5 });

		render(<CartSummary />);

		const negativeTotal = screen.getByText('Total a pagar: S/ -50.50');
		expect(negativeTotal).toBeInTheDocument();
	});

	test('matches snapshot', () => {
		const { asFragment } = render(<CartSummary />);
		expect(asFragment()).toMatchSnapshot();
	});
});
