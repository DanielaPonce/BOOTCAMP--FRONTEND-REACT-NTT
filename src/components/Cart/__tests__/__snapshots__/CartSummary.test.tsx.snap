// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartSummary Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="cart-summary"
  >
    <h3>
      Total a pagar: S/ 123.45
    </h3>
  </div>
</DocumentFragment>
`;
